## Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.
## Описание данных
`children` — количество детей в семье;  
`days_employed` — общий трудовой стаж в днях;  
`dob_years` — возраст клиента в годах;  
`education` — уровень образования клиента;  
`education_id` — идентификатор уровня образования;  
`family_status` — семейное положение;  
`family_status_id` — идентификатор семейного положения;  
`gender` — пол клиента;  
`income_type` — тип занятости;  
`debt` — имел ли задолженность по возврату кредитов;  
`total_income` — ежемесячный доход;  
`purpose` — цель получения кредита

## 1. Предобработка данных

In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Работаем с пропусками

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`.

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

Работа с аномальными значениями. Заменяем все отрицательные значения положительными.

In [ ]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выводим медианное значение трудового стажа `days_employed` в днях

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому не трогаем их. Тем более этот столбец не понадобится для исследования.

Выводим перечень уникальных значений столбца `children`

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

Удаляем аномальные значения

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Проверяем принятые изменения

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

Заполняем пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Проверяем принятые изменения

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Заменяем вещественный тип данных в столбце `total_income` на целочисленный

In [ ]:
data['total_income'] = data['total_income'].astype(int)

Обработаем неявные дубликаты в столбце `education`. Приведем их к нижнему регистру и проверим остальные столбцы

In [ ]:
data['education'] = data['education'].str.lower()

Смотрим на дубликаты и удаляем их

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведем перечень уникальных целей взятия кредита из столбца `purpose`

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.


In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

Проверим гипотезу о зависимости между количеством детей и возвратом кредита в срок

In [ ]:
children_debt = [] # создаем пустой список для результата вычисления
# вычисляем процент просроченных кредитов относительно выданных по каждому типу:
# сколько было всего просрочек по каждому типу (кол-ву детей) /
# сколько всего выдано кредитов по каждому типу (кол-ву детей)
children_debt = data.groupby('children')['debt'].sum() / data.groupby('children')['debt'].count() * 100
print(children_debt.sort_values(ascending = False)) # выводим результат и сортируем по убыванию

children
4    9.756098
2    9.454191
1    9.234609
3    8.181818
0    7.543822
5    0.000000
Name: debt, dtype: float64


Вывод:
Явной зависимости между количеством детей и возвратом кредита в срок нет.

Проверим гипотезу зависимости между семейным положением и возвратом кредита в срок

In [ ]:
family_status_debt = [] # создаем пустой список для результата вычисления
# вычисляем процент просроченных кредитов относительно выданных по каждому типу:
# сколько было всего просрочек по каждому типу (семейное положение) /
# сколько всего выдано кредитов по каждому типу (семейное положение)
family_status_debt = data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['debt'].count() * 100
print(family_status_debt.sort_values(ascending = False)) # выводим результат и сортируем по убыванию

family_status
Не женат / не замужем    9.763948
гражданский брак         9.313014
женат / замужем          7.560558
в разводе                7.064760
вдовец / вдова           6.624606
Name: debt, dtype: float64


Вывод:
Наибольший процент просрочек по кредиту имеют клиенты, которые не состоят или не состояли в официальных отношениях.

Проверим гипотезу о зависимости между уровнем дохода и возвратом кредита в срок?

In [ ]:
print('Медианное значение дохода:', data['total_income'].median())
print((data.groupby('total_income_category')['debt'].mean()*100).sort_values(ascending = False)) # выводим результат и сортируем по убыванию

Медианное значение дохода: 142594.0
total_income_category
E    9.090909
C    8.498210
A    8.000000
B    7.060231
D    6.017192
Name: debt, dtype: float64


Вывод: 
выборка по заемщикам, как с наименьшим доходом, так и с наибольшим крайне недостаточна для формирования однозначных выводов. Можно думать, что E-категория самая низкооплачиваемая не может корректно распоряжаться деньгами и «влезает» в кредиты, но тогда неясно почему у А-категории также достаточно высокий процент задолженности. В общем вывод по группам А и Е неочевиден, мало данных.
Если же взять для анализа только группы В, С и D, то по полученным данным не наблюдаем зависимости "те, кто получает больше ЗП, лучше возвращает кредит». В целом это логично, так как материальный достаток - не единственная переменная, которая влияет на качество кредитной истории.

Проверим гипотезу как разные цели кредита влияют на его возврат в срок

In [ ]:
data['purpose_category'].unique() # Смотрим уникальные значения по столбцу 
purpose_category_debt = [] # создаем пустой список для результата вычисления
# вычисляем процент просроченных кредитов относительно выданных по каждому типу:
purpose_category_debt = data.groupby('purpose_category')['debt'].sum() / data.groupby('purpose_category')['debt'].count() * 100
print(purpose_category_debt.sort_values(ascending = False)) # выводим результат и сортируем по убыванию

purpose_category
операции с автомобилем      9.347978
получение образования       9.252758
проведение свадьбы          7.911803
операции с недвижимостью    7.255139
Name: debt, dtype: float64


Вывод:
Наибольший процент просрочек по кредиту на автокредите и на получении образования. Меньший процент просрочек при кредитовании недвижимости и свадеб.

## Итог  
***Задача:***  
Кредитным отделом банка поставлена задача - проверить зависимость семейного положения и количества детей клиента на факт
погашения кредита в срок.  
В качестве исходных данных предоставлена статистика о платежеспособности клиента.  
Результат исследования будет учтен при построении модели кредитного скоринга - системы способной оценить вероятность возврата кредита банку потенциальным заёмщиком.
  
***Гипотезы:***  
Для решения задачи было поставлено несколько гипотез:  
1 Возврат кредита в срок зависит от количества детей
2 Возврат кредита в срок зависит от семейного положения
3 Возврат кредита в срок зависит от уровня дохода
4 Возврат кредита в срок зависит от цели кредита
  
***Результаты:***  
  
*Гипотеза 1. Возврат кредита в срок зависит от количества детей.*  
а. С 4 детьми доля просроченных кредитов составляет   9.75%  
б. С 2 детьми доля просроченных кредитов составляет   9.45%  
в. С 1 ребенком доля просроченных кредитов составляет 9.23%  
г. С 3 детьми доля просроченных кредитов составляет   8.18%  
д. Для людей без детей доля просроченных кредитов составляет 7.54%  
е. С 5 детьми доля просроченных кредитов составляет    0.00%

Гипотеза 1 не подтвердилась, явной зависимости не наблюдается.

*Гипотеза 2. Возврат кредита в срок зависит от семейного положения.*  
а. По группе "Не женат / не замужем" доля просроченных кредитов составляет 9.76%
б. По группе "гражданский брак" доля просроченных кредитов составляет 9.31%
в. По группе "женат / замужем"доля просроченных кредитов составляет 7.56%
г. По группе "в разводе" доля просроченных кредитов составляет 7.06%
д. По группе "вдовец / вдова" доля просроченных кредитов составляет 6.62%
  
Гипотеза 2 подтверждена. Люди, которые состояли в официальных отношениях чаще отдают кредит в срок относительно  
людей, которые в них не состояли.
  
*Гипотеза 3. Возврат кредита в срок зависит от уровня дохода.*  
а. У людей с доходом в интервале от 0...30 000 доля просрочек по кредиту составляет 9.09%  
б. У людей с доходом в интервале от 50 001...200 000 доля просрочек по кредиту составляет 8.49%  
в. У людей с доходом в интервале от 1 000 001 и выше доля просрочек по кредиту составляет 8.00%  
г. У людей с доходом в интервале от 200 001...1 000 000 доля просрочек по кредиту составляет 7.06%  
д. У людей с доходом в интервале от 30 001...50 000 доля просрочек по кредиту составляет 6.01%  
  
Гипотеза 3 частично подтвердилась. У людей с доходом до 30 000 процент просрочек выше.  
Рекомендации: провести анализ с большим количеством выборки по уровню дохода.  
  
*Гипотеза 4. Возврат кредита в срок зависит от цели кредита.*  
а. В категории "операции с автомобилем" доля просроченных кредитов составляет 9.34%  
б. В категории "получение образования" доля просроченных кредитов составляет 9.25%  
в. В категории "проведение свадьбы" доля просроченных кредитов составляет 7.91%  
г. В категории "операции с недвижимостью" доля просроченных кредитов составляет 7.25%  
  
Гипотеза 4 подтверждена. Наименьший процент просрочки имеют кредиты с целями "операции с недвижимостью" и "проведение свадьбы".  
  
**Общий вывод**  
  
а. Для построения модели кредитного скоринга следует опираться на гипотезы 2 (Возврат кредита в срок зависит от  
семейного положения) и 4 (возврат кредита в срок зависит от цели кредита)  
б. Для гипотезы 3 (возврат кредита в срок зависит от уровня дохода) необходимо повторить анализ с большим количеством выборки.